http/src/main/java/io/micronaut/http/filter/FilterRunner.java

```diff
@@ -0,0 +1,966 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.filter;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.async.publisher.Publishers;
+import io.micronaut.core.execution.CompletableFutureExecutionFlow;
+import io.micronaut.core.execution.ExecutionFlow;
+import io.micronaut.core.execution.ImperativeExecutionFlow;
+import io.micronaut.core.order.OrderUtil;
+import io.micronaut.core.order.Ordered;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.type.Executable;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.HttpResponse;
+import io.micronaut.http.MutableHttpRequest;
+import io.micronaut.http.MutableHttpResponse;
+import io.micronaut.http.reactive.execution.ReactiveExecutionFlow;
+import io.micronaut.inject.ExecutableMethod;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.CorePublisher;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Mono;
+import reactor.util.context.Context;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+/**
+ * @implNote Legacy filters had a strict execution flow: filter1->chain.proceed->filter2->chain.proceed...
+ * This leads to deep call stacks and requires use of reactive flows. The new filter API is more
+ * flexible. Filters do not have to accept a chain they call proceed on, they do not have to use
+ * reactive types, and so on.<br>
+ * {@link FilterRunner} takes advantage of this flexibility to optimize execution of filters. Most
+ * importantly, if a filter returns an immediate value (as opposed to a reactive flow),
+ * {@link FilterRunner} can execute filters <i>sequentially</i>, instead of in a recursive fashion.<br>
+ * The implementation of this is inspired by kotlin coroutines. Filter execution essentially
+ * happens in a loop ({@link #workRequest()} and {@link #workResponse()}), but the loop counter is
+ * an instance variable ({@link #index}), and the loop can sometimes <i>suspend</i>, e.g. when a
+ * filter returns a reactive flow instead of an immediate value. When suspension happens, the loop
+ * exits early, and {@link #workRequest()} (or {@link #workResponse()}) will be called again on
+ * unsuspend (e.g. when the reactive flow completes).
+ *
+ * @since 4.0.0
+ * @author Jonas Konrad
+ */
+@Internal
+public class FilterRunner {
```

Comment:
It feels to me that some of the code can be extracted into small methods to explain what it does. There is so much work with fields, so it's hard to read.

Reply:
It looks a bit suspicious to me that some of the methods can be executed on a different thread (If I understand correctly) `executeOn.execute``terminalFlow.onComplete` with the class fields being modified, if this is a case, I would have written this as one flow carrying some data structure along the stream.

Reply:
Most of the time, there is only one thread that modifies the class fields. e.g. when there is an ExecuteOn, it suspends the work on the original thread, and resumes on the new thread.

There are some exceptions to this, e.g. with a continuation (or legacy filter), the user could return from the filter (or complete the publisher returned by the filter) before the downstream `proceed` is done. These cases are handled by the concurrency structures in the SuspensionPoint subclasses